src/main/java/frc/robot/commands/DriveCommand.java

```diff
@@ -28,6 +31,9 @@ public void initialize() {
   // Called every time the scheduler runs while the command is scheduled.
   @Override
   public void execute() {
+    drive_subsystem.drive(
+        Math.min(m_container.mainController.getMoveRequest(), m_container.mainController.getSpeedLimiter()),
```

Comment:
I'm pretty sure you want to be using a map here, not a min.

Although, it's probably architecturally better to just deal with all the input processing (speed limiting, deadzone mapping, input curving) in the DriveController class and just pass the desired move request straight to the drive subsystem.
So you probably don't need the `getSpeedLimiter` method in the ControllerBase.

I figure you're probably using it because we used it in previous years. However, I'm pretty sure that's a holdover from when we were doing Iterative robots and not Command-Based robots. To me it seems better to separate the input processing logic from the driving logic.

Reply:
In fact, now that I look at the code in ControllerBase, you have MOVE_CURVE and TURN_CURVE constants (which should probably be private), but you never use them.

Reply:
May I have some clarification on the first part(specifically, dealing with all input processing in DriveController)?  I'm a little confused.  Wouldn't the Command be the best place to do that?

Reply:
Well, I may be mistaken, but to me it makes the most sense that the DriveController determines what speed the user wants to move at, and the DriveCommand tells to DriveSubsystem to drive at that speed.
The DriveController does some processing (dead zones, input curves, speed limiting) to better interpret what the user wants, but interpreting the user’s desired speed is the DriveController’s job.

This is especially relevant for different types of controller. What processing you do for an Xbox controller is likely different to the processing you do for a Joystick. However, you want one single DriveCommand to be in charge of driving, no matter where the user input is coming from.

Reply:
I concur with Jordan that there should be a logical separation of responsibility between the layers.  There are some available options for slicing within the stack, but the ordering is pretty fixed.
Fixed order of operations:

1. read raw user inputs
1. condition inputs (deadzone, curves, etc) to make values linear and normalized (-1,0,1)
1. apply additional constraints/conditions like speed scaling
1. apply closed-loop, fly-by-wire corrections (auto stability, etc) at the move/turn level
1. translate move/turn requests into drive-specific requests (move/turn -> left/right)
1. apply additional conditioning to drive-specific requests (motor velocity PID)

Clearly, 1-2 belong in the DriveController, and 5-6 belong in the DriveSubsystem (6 actually lives in the Falcons this year).  I would argue that 4 would typically end up in the DriveSubsystem, but it could be refactored into a separate layer if it has different interactions between auton and teleop modes.  Note that none of these steps has a specific affinity to DriveCommand.

So the question comes down to where to put 3.  The extreme ends of the spectrum are:

- Bake it into any one of the layers: Controller, Command, Subsystem, based on where you *think* you want to restrict its scope of influence.  i.e. If you're sure you only ever want it to affect a particular Controller, then put it there.  If you're sure you want to apply to every possible mode of control, then put it in the subsystem.
- Create a ConditionedDriveSubsystem that manages the inputs from the DriveCommand and then passes them along to the lower-level DriveSubsystem.

Breaking it out as an independent layer gives you the most flexibility, but also the most overhead.  Given the fact that we'll likely need flexibility vs performance, I would recommend going the ConditionedDriveSubsystem route.  This allows you to create button mappings and commands that can adjust the speed scaler and then the speed scaler conditions the inputs to the DriveSubsystem.

Bottom line is that it appears we're keen to keep the Commands as lightweight as possible.  This increases our ability to use InstantCommands: https://docs.wpilib.org/en/latest/docs/software/commandbased/convenience-features.html